docs/posts/duckdb-for-rag/index.qmd

```diff
@@ -0,0 +1,512 @@
+```
+
+Let's jump into the code!
+
+## RAG over Ibis documentation
+
+We'll use the Ibis documentation as our knowledge base.
+
+### Setup code
+
+First we need to install and import the necessary packages:
+
+```bash
+pip install 'ibis-framework[duckdb]' openai python-dotenv
+```
+
+```{python}
+import os  # <1>
+import ibis
+import shutil
+import subprocess
+
+from openai import OpenAI
+from dotenv import load_dotenv  # <1>
+```
+
+1. Import the libraries we'll use.
+
+We need to set the `OPENAI_API_KEY` environment variable to our OpenAI API key.
+Do this by creating a `.env` file in the same directory as the code and adding
+the following line:
+
+```txt
+OPENAI_API_KEY=<your-api-key>  # <1>
+```
+
+1. Replace with your OpenAI API key.
+
+Then load the environment variables:
+
+```{python}
+load_dotenv()  # <1>
+```
+
+1. Load environment variables from the `.env` file.
+
+Now we're ready to make OpenAI calls.
+
+Let's also setup Ibis for interactive use:
+
+```{python}
+ibis.options.interactive = True  # <1>
+```
+
+1. Turn on Ibis interactive mode.
+
+### Getting the data into DuckDB
+
+We'll use a function to clone a given repository:
+
+```{python}
+def clone_repo(repo_url):  # <1>
+    subprocess.run(["git", "clone", "--depth", "1", repo_url], check=False)  # <1>
+```
+
+1. Define a function to clone a repository.
+
+Let's clone `ibis` and prepare to ingest all the Quarto markdown files into a
+DuckDB database called `rag.ddb`:
+
+```{python}
+org_url = "https://github.com/ibis-project"  # <1>
+repo_name = "ibis"  # <1>
+
+shutil.rmtree(repo_name, ignore_errors=True)  # <2>
+clone_repo(f"{org_url}/{repo_name}")  # <3>
+
+if os.path.exists("rag.ddb"):  # <4>
+    os.remove("rag.ddb")
+if os.path.exists("rag.ddb.wal"):
+    os.remove("rag.ddb.wal")  # <4>
+
+rag_con = ibis.connect("duckdb://rag.ddb")  # <5>
+rag_con.list_tables()  # <6>
+```
+
+1. Setup the repository to clone.
+2. Remove the repository if it exists to ensure a fresh clone.
+3. Clone the repository.
+4. Remove the DuckDB database if it exists to ensure a fresh database.
+5. Connect to the DuckDB database.
+6. List the tables in the database, showing it is empty.
+
+We'll walk through all the cloned files. For `.qmd` files, read the contents and
+insert them into the `ibis` table:
+
+```{python}
+for root, _, files in os.walk(repo_name):  # <1>
+    for file in files:  # <2>
+        if file.endswith(".qmd"):  # <3>
+            with open(os.path.join(root, file)) as f:  # <4>
+                contents = f.read()  # <5>
+
+                data = {  # <6>
+                    "filepath": [os.path.join(root, file)],
+                    "contents": [contents],
+                }  # <6>
+
+                t = ibis.memtable(data)  # <7>
+
+                if repo_name not in rag_con.list_tables():  # <8>
+                    rag_con.create_table(repo_name, t)  # <8>
+                else:  # <9>
+                    rag_con.insert(repo_name, t)  # <9>
```

Comment:
```suggestion
            data = {  # <6>
                "filepath": [os.path.join(root, file)],
                "contents": [contents],
            }  # <6>

            t = ibis.memtable(data)  # <7>

            if repo_name not in rag_con.list_tables():  # <8>
                rag_con.create_table(repo_name, t)  # <8>
            else:  # <9>
                rag_con.insert(repo_name, t)  # <9>
```

Should be able to dedent all these lines and only keep the file open for as long as necessary, and reduce nesting a bit